### PR TITLE
[PM-4165] Collect all Stripe invoices for Billing history page

### PR DESCRIPTION
**Pull request by amorask-bitwarden**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

[PM-4165](https://bitwarden.atlassian.net/browse/PM-4165)

Currently, for the Billing history page, we only collect the first 50 invoices a customer has. This PR updates that logic to collect all the invoices using the optional [`starting_after`](https://stripe.com/docs/api/invoices/list#list_invoices-starting_after) property in the Stripe List Invoices endpoint.

## Code changes

* **StripePaymentService.cs:** Retrieve invoices from Stripe's List Invoices endpoint until there are none left to retrieve. Use a 100 limit to limit this to as few network requests as possible.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

[PM-4165]: https://bitwarden.atlassian.net/browse/PM-4165?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by amorask-bitwarden** [comment]
The method changed here is private and the calling method is here: https://github.com/bitwarden/server/blob/4ff5d3951ba91723103852ff686560ca591ffb7a/src/Core/Services/Implementations/StripePaymentService.cs#L1523-L1535

In order to efficiently add any unit tests, this method would have to effectively be broken up less we have to mock every response required to build up the entire `BillingInfo` object.

**Comment by eliykat** [comment]
@cturnbull-bitwarden could you please review this for @bitwarden/tech-leads given that it's touching `StripePaymentService`?

Also, can you move `StripePaymentService` into the Billing folder so that you have code ownership? (not necessarily this PR, just flagging it)

File: src/Core/Services/Implementations/StripePaymentService.cs [hunk 1]
```diff
@@ -1850,15 +1850,35 @@ private async Task<IEnumerable<BillingInfo.BillingInvoice>> GetBillingInvoicesAs
             return null;
         }
 
-        var invoices = await _stripeAdapter.InvoiceListAsync(new Stripe.InvoiceListOptions
+        return (await Collect())
+            .Where(invoice => invoice.Status != "void" && invoice.Status != "draft")
+            .OrderByDescending(invoice => invoice.Created)
+            .Select(invoice => new BillingInfo.BillingInvoice(invoice));
+
+        async Task<List<Stripe.Invoice>> Collect()
```
[the remark this answers is not in the document]
Comment (amorask-bitwarden) [reply src/Core/Services/Implementations/StripePaymentService.cs]: Exception handling added here: https://github.com/bitwarden/server/pull/3376/commits/2fcf295af883a54eb71525bdcce6855cbe55c2a8

File: src/Core/Services/Implementations/StripePaymentService.cs [hunk 2]
```diff
@@ -1850,15 +1850,35 @@ private async Task<IEnumerable<BillingInfo.BillingInvoice>> GetBillingInvoicesAs
...
+            {
+                var invoiceListOptions = new Stripe.InvoiceListOptions { Customer = customer.Id, Limit = 100 };
+
+                if (collected.Any())
+                {
+                    var lastInvoiceId = collected.Last().Id;
+                    invoiceListOptions.StartingAfter = lastInvoiceId;
+                }
+
+                var invoices = await _stripeAdapter.InvoiceListAsync(invoiceListOptions);
```
[the remark this answers is not in the document]
Comment (amorask-bitwarden) [reply src/Core/Services/Implementations/StripePaymentService.cs]: Exception handling added here: https://github.com/bitwarden/server/pull/3376/commits/2fcf295af883a54eb71525bdcce6855cbe55c2a8

File: src/Core/Services/Implementations/StripePaymentService.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (amorask-bitwarden) [reply src/Core/Services/Implementations/StripePaymentService.cs]: @cturnbull-bitwarden I actually found another instance of this already being done with subscriptions, which I've emulated for my change as well:
https://github.com/bitwarden/server/blob/2fcf295af883a54eb71525bdcce6855cbe55c2a8/src/Core/Services/Implementations/StripeAdapter.cs#L197-L211

It just uses a derived Stripe options class to add a `SelectAll` property and that property determines whether or not to use the Stripe service's `ListAutoPagingAsync` method. Reference: https://stripe.com/docs/api/pagination/auto

**Review by cturnbull-bitwarden — APPROVED** [review]
Nice work!

**Review by cturnbull-bitwarden — APPROVED** [review]

**Pull request closed without merging**